pandasai/llm/gpt4all.py

```diff
@@ -0,0 +1,181 @@
+    it might not work as chatGPT when it comes to performance, hence for now users using
+    this module have to tune the existing prompts to get better results.
+    """
+
+    temp: Optional[float] = 0
+    top_p: Optional[float] = 0.1
+    top_k: Optional[int] = 40
+    n_batch: Optional[int] = 8
+    n_threads: Optional[int] = 4
+    n_predict: Optional[int] = 256
+    max_tokens: Optional[int] = 200
+    repeat_last_n: Optional[int] = 64
+    repeat_penalty: Optional[float] = 1.18
+
+    _model_repo_url = "https://gpt4all.io/models/models.json"
+    _supported_models = [
+        metadata["filename"] for metadata in requests.get(_model_repo_url).json()
+    ]
+
+    def __init__(
+        self,
+        model_name: str,
+        model_folder_path: Optional[str] = None,
+        allow_download: Optional[bool] = False,
+        n_threads: Optional[int] = 4,
+        download_chunk_size: Optional[int] = 8912,
+        **kwargs,
+    ) -> None:
+        """
+        GPT4All client for using Pandas AI
+        Args:
+            model_name: The name of the model.
+            model_folder_path: The folder inside the model weights are present
+            allow_download: If True will trigger download the specified model
+            n_threads: Number of CPU threads to be used while running the model
+            download_chunk_size: The chunk size set for downloading the model
+        """
+        self.model_name = (
+            f"{model_name}.bin" if not model_name.endswith(".bin") else model_name
+        )
+        self.chunk_size = download_chunk_size
+
+        _assert_msg = (
+            "Model not Found. Please check the list of supported models:",
+            self._supported_models,
+        )
+        assert self.model_name in self._supported_models, _assert_msg
+
+        # automatically download inside that folder
+        _default_download_path = os.path.join(
+            os.path.expanduser("~"), ".local/share/nomic.ai/GPT4All/"
+        )
+
+        self.model_folder_path = (
+            _default_download_path if model_folder_path is None else model_folder_path
+        )
+
+        # automatically create the default folder and download the model
+
+        if not os.path.exists(self.model_folder_path) and allow_download:
+            os.mkdir(self.model_folder_path)
+
+        if allow_download:
+            self._auto_download()
+            print(f"Model {model_name} saved as: {self.model_folder_path}")
+
+        n_threads = self.n_threads if n_threads is None else n_threads
+        try:
+            from gpt4all import GPT4All
+
+            self.gpt4all_model = GPT4All(
+                model_name=self.model_name,
+                model_path=self.model_folder_path,
+                n_threads=n_threads,
+            )
+        except ImportError:
+            raise ImportError(
+                "Unable to import gpt4all python package "
+                "Please install it with `pip install -U gpt4all`"
+            )
```

Comment:
```suggestion
        self.gpt4all_model = GPT4All(
            model_name=self.model_name,
            model_path=self.model_folder_path,
            n_threads=n_threads,
        )
```
Imports shouldn't be mixed with code logic except some extraordinary cases. (see [PEP-08 Imports](https://peps.python.org/pep-0008/#imports)).

Reply:
Yeah, I saw this one. Hence the try-exception for the import was added earlier. Makes sense.